### PR TITLE
[pdex] Add statev2 code base and refactoring state interface by state…

### PR DESCRIPTION
**Pull request by tin-incognito**
*Add*
1. state v2 basic structure

*Change*
1. Add interface reader for PDE state

### CODE REVIEW

File: blockchain/pdex/interface.go [hunk 1]
```diff
@@ -14,9 +10,13 @@ type State interface {
...
-	PoolPairs() map[string]*rawdbv2.PDEPoolForPair
+	Reader() StateReader
+}
+
+type StateReader interface {
+	WaitingContributions() map[string]interface{}
+	DeletedWaitingContributions() map[string]interface{}
+	PoolPairs() map[string]interface{}
 	Shares() map[string]uint64
 	TradingFees() map[string]uint64
```
Comment (qcuong98) [inline blockchain/pdex/interface.go:21]: Shares() and TradingFees() are belonging to the PoolPair State

File: blockchain/pdex/state_base.go [hunk 2]
```diff
@@ -55,15 +51,15 @@ func (s *stateBase) GetDiff(compareState State) (State, error) {
 	panic("Implement this fucntion")
 }
 
-func (s *stateBase) WaitingContributions() map[string]*rawdbv2.PDEContribution {
+func (s *stateBase) WaitingContributions() map[string]interface{} {
 	panic("Implement this fucntion")
```
Comment (qcuong98) [inline blockchain/pdex/state_base.go:55]: Typo (fucntion)

**Review by dat-incognito-org — APPROVED** [review]

File: blockchain/pdex/state_v2.go [hunk 3]
```diff
@@ -12,47 +11,50 @@ type stateV2 struct {
...
+	stakers          map[string]uint64 // nfst -> amount staking
+	currentStakingID uint64
+}
+
+type Order struct {
+	tick            int64
+	tokenBuyID      string
+	tokenBuyAmount  uint64
+	tokenSellAmount uint64
+	ota             string
```
Comment (qcuong98) [inline blockchain/pdex/state_v2.go:33]: The OTAs using for storage are composed of two strings (OTAAddressStr, TxRandomStr). Not just one

File: blockchain/pdex/state_v2.go [hunk 4]
```diff
@@ -12,47 +11,50 @@ type stateV2 struct {
...
+	fee             uint64
+	txReqID         string
```
Comment (qcuong98) [inline blockchain/pdex/state_v2.go:35]: change to reqTxID for unifying with other tx types

File: blockchain/pdex/state_v2.go [hunk 5]
```diff
@@ -12,47 +11,50 @@ type stateV2 struct {
...
-	token2Amount          uint64
-	minPrice              float64 // Compare price between token 1 and token 2
-	maxPrice              float64
-	txReqID               common.Hash
+	otaRefund   string // refund contributed token
+	otaReceive  string // receive nfct
+	tokenID     string
+	tokenAmount uint64
+	amplifier   uint // only set for the first contribution
+	txReqID     string
```
Comment (qcuong98) [inline blockchain/pdex/state_v2.go:44]: Between a token pair (ex. ETH/PRV). There is a chance that our protocol stores lots of AMM pool pairs (difference in P_0 and amplifier configuration). The contribution state should also store the poolPairID of the desired pool that the LP wants to contribute.

File: blockchain/pdex/interface.go (hunk 1 above)
Comment (tin-incognito) [reply blockchain/pdex/interface.go:21]: These functions are for state version 1

File: blockchain/pdex/state_v2.go (hunk 3 above)
Comment (tin-incognito) [reply blockchain/pdex/state_v2.go:33]: ok

File: blockchain/pdex/state_v2.go (hunk 5 above)
Comment (tin-incognito) [reply blockchain/pdex/state_v2.go:44]: ok

File: blockchain/pdex/state_v2.go (hunk 4 above)
Comment (tin-incognito) [reply blockchain/pdex/state_v2.go:35]: ok

**Review by cuong-incognito — APPROVED** [review]
LGTM!

File: blockchain/pdex/interface.go (hunk 1 above)
Comment (0xankylosaurus) [reply blockchain/pdex/interface.go:21]: It's okay to change a concrete type to an interface but you're using `map[string]interface{}`, so when we want to use these properties' values, we have to do assertion, correct?
If so, why don't we define explicit interfaces for these?

**Review by 0xankylosaurus — APPROVED** [review]

**Pull request merged**